{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8.5"
    },
    "colab": {
      "name": "Assignment1 _AditiGopalan.ipynb",
      "provenance": [],
      "collapsed_sections": []
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Ho08FGbzegBv"
      },
      "source": [
        "This is Assignment 1. Deadline for submission is October, 3, 11pm EST. Please send your solutions in a gzipped archive to dmm2017@med.cornell.edu, with copy to chb4004@med.cornell.edu. No extensions will be granted.\n",
        "\n",
        "Your archive should contain this ipynb file with the code you wrote. However, if you are not comfortable with python, you can write in any other programming language, but provide instructions how to run/compile your code. "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ydPZ5qvNegB1"
      },
      "source": [
        "Assignment 1, Aditi Gopalan\n",
        "\n",
        "Task 1:\n",
        "====\n",
        "\n",
        "Your task is to find the position of an insertion in a given genome. The length of each insertion is unknown. You will be given five input files. Random sequence contain A,C,G,T characters and probabilities of each character are equal.\n",
        "\n",
        "01.fasta - a real genomic sequence with a random sequence inserted in a random place.  \n",
        "\n",
        "02.fasta - a heavily mutated real genomic sequence with a random sequence inserted in a random place\n",
        "\n",
        "03.fasta - a real genomic sequence with real genomic sequence from a completely different organism inserted in a random place \n",
        "\n",
        "04.fasta - a real genomic sequence with real genomic sequence from a completely different organism inserted in a random place\n",
        "\n",
        "05.fasta - a randomly generated sequence with a real genomic sequence inserted in a random place\n",
        "\n",
        "**Output format**\n",
        "\n",
        "For each input file output two positions - start and end position of inserted sequence.\n",
        "\n",
        "**Grading**\n",
        "\n",
        "Your solution will be graded based on a secret formula that penalizes the solution if your output start and end positions are far away from the truth start and end.\n",
        "Only answers will be graded, but you will have to provide code that you used to get this solution."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "9FEUCS-4fsWk"
      },
      "source": [
        "\n",
        " **OUTPUT:**\n",
        "\n",
        "I was able to identify areas of insertion on the gene: \n",
        "\n",
        "fasta1: 17800 22000\n",
        "\n",
        "fasta2: 1490250 1563700 \n",
        "\n",
        "fasta3: 4900 6380\n",
        "\n",
        "fasta4: 362870 410000\n",
        "\n",
        "fasta5: 1177000 1235123 "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "PSjUFZekegB3"
      },
      "source": [
        "import pandas as pd\n",
        "from Bio import SeqIO\n",
        "import operator\n",
        "import numpy as np\n",
        "import matplotlib\n",
        "from matplotlib import pyplot as plt\n",
        "\n",
        "'''\n",
        "Collaborated with Max Chao \n",
        "\n",
        "My approach is to check the GC content on each fasta file because I assume the insertions will show a different pattern of GC content. This will help to identify regions of insertion. \n",
        "\n",
        "For fasta files 2 and 5 I determined the kmer position in order to narrow down the region of insertion. \n",
        "\n",
        "OUTPUT:\n",
        "I was able to identify areas of insertion on the gene:\n",
        "fasta1: 17800 22000\n",
        "fasta2: 1490250 1563700 \n",
        "fasta3: 4900 6380\n",
        "fasta4: 362870 410000\n",
        "fasta5: 1177000 1235123\n",
        "\n",
        "'''\n",
        "def get_gc(seq): #counts GC content\n",
        "\tcount_c = 0\n",
        "\tcount_g = 0\n",
        "\tfor i in seq:\n",
        "\t\tif i == \"C\":\n",
        "\t\t\tcount_c +=  1\n",
        "\t\tif i == \"G\":\n",
        "\t\t\tcount_g += 1\n",
        "\ttotal = count_c + count_g\n",
        "\treturn total\n",
        "\n",
        "\n",
        "fasta_sequences = SeqIO.parse(open(\"/mnt/c/ds/01.fasta\"),'fasta') #parsing a fasta file\n",
        "with open(\"/mnt/c/ds/01.fasta\") as fasta_file:\n",
        "\tfor fasta in fasta_sequences:\n",
        "\t\tname, seq = fasta.id, str(fasta.seq)\n",
        "\t\tprint(\"Len\", len(seq))\n",
        "\t\tbin_size = len(seq)//100 #Splitting the sequence into 100 bins of equal nucleotides\n",
        "\t\tCG_dict = {} #Dictionary that stores the GC composition of each bin\n",
        "\t\tfor n in range(0,100):\n",
        "\t\t\tif n == 0:\n",
        "\t\t\t\tbin_n = seq[0:bin_size]\n",
        "\t\t\t\tgc = get_gc(bin_n)\n",
        "\n",
        "\t\t\telif n == bin_size-1:\n",
        "\t\t\t\tbin_n = seq[(bin_size*n):]\n",
        "\t\t\t\tgc = get_gc(bin_n)\n",
        "\n",
        "\t\t\telse:\n",
        "\t\t\t\tbin_n = seq[bin_size*n:bin_size*(n+1)]\n",
        "\t\t    gc = get_gc(bin_n)\n",
        "\t\t\tgc_comp = gc/len(bin_n)\n",
        "\t\t\tCG_dict[n] = gc_comp\n",
        "\t\tplt.bar(*zip(*CG_dict.items())) #Plotting GC composition to identify areas of irregularity\n",
        "\t\tplt.savefig('fasta1.png')\n",
        "\n",
        "\n",
        "\t\tinsert_pos = []\n",
        "\t\tfor k, v in CG_dict.items():\n",
        "\t\t\tif v> 0.48:#Setting threshold value from region of irregularity in the graph\n",
        "\t\t\t\tinsert_pos.append(k)\n",
        "\t\tstart = min(insert_pos)\n",
        "\t\tend = max(insert_pos)\n",
        "\t\tprint(start, end)\n",
        "\t\ts = start*bin_size\n",
        "\t\te = end*bin_size\n",
        "\t\tprint(\"The insertion is approximately between\", (s,e) )\n",
        "\n",
        "\t\t\n",
        "'''\n",
        "Since we cannot tell the areas of insertion from just GC composition for fasta2 and fasta5, I decided to try a kmer approach to identify these regions\n",
        "using kmers, for fast 2 and fast 5\n",
        "\n",
        "Note that this comment section separates two differing python files of code\n",
        "'''\n",
        "import pandas as pd\n",
        "from Bio import SeqIO\n",
        "import operator\n",
        "import numpy as np\n",
        "import matplotlib\n",
        "from matplotlib import pyplot as plt\n",
        "\n",
        "def find_kmer(name, seq): #function counts kmers in a given sequence\n",
        "\tprint(\"%s\\n################################\" %name)\n",
        "\tkmers = {} #this is a dictionary that will record kmers and their frequencies\n",
        "\tpos = {} #this is a dictionary of all the kmer positions\n",
        "\tk = 7 #this is kmer size\n",
        "\tfor i in range(len(seq) - k + 1): #counting kmers\n",
        "\t\tkmer = seq[i:i+k]\n",
        "\t\tpos_k = [i, i+k]\n",
        "\t\tif kmer in kmers:\n",
        "\t\t\tkmers[kmer] += 1\n",
        "\t\telse:\n",
        "\t\t\tkmers[kmer] = 1\n",
        "\t\tif pos_k in pos:\n",
        "\t\t\tpos[kmer] += 1\n",
        "\t\telse:\n",
        "\t\t\tpos[kmer] = 1\n",
        "\tdata_items = kmers.items()\n",
        "\tdata_list = list(data_items)\n",
        "\tdf = pd.DataFrame(data_list)\n",
        "\tprint(df)\n",
        "\n",
        "\tdata_items1 = pos.items()\n",
        "\tdata_list1 = list(data_items1)\n",
        "\tdf1 = pd.DataFrame(data_list1)\n",
        "\tprint(\"df1\", df1)\n",
        "\n",
        "\tinsert_pos = []\n",
        "\tfor k, v in kmers.items():\n",
        "\t\tif  v <= 10:\n",
        "\t\t\tfor k2, v2 in pos.items():\n",
        "\t\t\t\t   c2 = pos['k']\n",
        "\t\t\t       insert_pos.append(c2)\n",
        "\n",
        "\tstart = min(insert_pos)\n",
        "\tend = max(insert_pos)\n",
        "\tprint(\"The insertion is approximately between\", (start,end) )\n",
        "\n",
        "fasta_sequences = SeqIO.parse(open(\"/mnt/c/ds/02.fasta\"),'fasta') #parsing a fasta file\n",
        "with open(\"/mnt/c/ds/02.fasta\") as fasta_file:\n",
        "\tfor fasta in fasta_sequences:\n",
        "\t\tname, sequence = fasta.id, str(fasta.seq)\n",
        "\t\tfind_kmer(name, sequence)\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "p_sNKDR-egB4"
      },
      "source": [
        "Task 2:\n",
        "====\n",
        "Suppose that you identified the outbreak of unknown virus in a distant village. Villagers want to know who was the first person, who get the virus. Unfortunately, virus has a long and variable incubation period and it can transmit between people even during incubation. You sampled and sequenced 100 randomly selected patients from the village and assembled viral RNA. Your task is to order these people according to their infection time.\n",
        "\n",
        "When an infected person touches a healthy person, he/she gets the virus, during transmission virus mutates but it's DNA doesn't change afterwards. It can get some SNPs (single nucleotide polymorphisms) and indels during the mutation procedure. In our simulations we uniformly and randomly introduce 0-3 deletions and 1-5 insertions per transmission.\n",
        "\n",
        "Your solution will be rated according to the number of correct strictly ordered pairs in our simulations. We simulated data according to some rooted tree with zero patient being in the root. It means that there is no reinfection.\n",
        "\n",
        "You should get your answer in some kind of automated way. Please, provide your code for computing global alignment and final arrangement. Briefly describe your ideas in a write-up or separate notebook cell.\n",
        "\n",
        "**Input format:**\n",
        "\n",
        "Input file contains 100 DNA sequences, one per line. In i-th line there is a assembled virus from i-th patient. \n",
        "\n",
        "**Output format:**\n",
        "\n",
        "Please provide your answer in a separate file. Your file should contain 100 distinct integer numbers, one per line. Number $n$ on i-th line means that you think that n-th person was the i-th person to get the virus.\n",
        "  \n",
        "Answer:  \n",
        "\n",
        "2  \n",
        "\n",
        "3  \n",
        "\n",
        "1  \n",
        "\n",
        "means that transmission history is [2->3, 2->1] or [2->3->1] \n",
        "\n",
        "Input file can be found on the Google Disk.\n",
        "\n",
        "**Score distribution for this task:**\n",
        "\n",
        "1. 50% - correct ordering\n",
        "2. 30% - code quality\n",
        "3. 20% - idea correctness"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "LqoEIVQ8egB5"
      },
      "source": [
        "import itertools\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import itertools\n",
        "import regex as re\n",
        "import os\n",
        "import glob\n",
        "import pandas as pd\n",
        "os.chdir(\"/mnt/c/ds\")\n",
        "\n",
        "\"\"\"\n",
        "Collaborated with Yohyoh Wang and Phoebe Fei \n",
        "\n",
        "I first sorted the input sequences based on length into five bins of 20 sequences each. I performed pairwise global alignment on each of these bins to compute the alignments and score.\n",
        "Once all the bins were aligned, I counted the number of mutatational events. If consecutive, I counted them as one event. Then, I sorted pairs based on ascending order of scores to obtain the order of sequences.\n",
        "The sequences with the maximum score were the most mutated or \"newest\" sequences\n",
        "\"\"\"\n",
        "\n",
        "def aligner(s1, s2):\n",
        "\t#scores\n",
        "\tmismatch = 1\n",
        "\tmatch = 0\n",
        "\tindel = 1\n",
        "\n",
        "\t#Initializing the matrix\n",
        "\tfor a in range(1, m):\n",
        "\t\tmat[a, 0] = a\n",
        "\t\ttrace[a,0] = 1\n",
        "\tfor b in range(1, n):\n",
        "\t\tmat[0, b] = b\n",
        "\t\ttrace[0, b] = 2\n",
        "\n",
        "\t#Scoring matrix based on match/mismatch\n",
        "\tfor i in range(1, m):\n",
        "\t\tfor j in range(1, n):\n",
        "\n",
        "\t\t\tif seq1[i-1] == seq2[j-1]:\n",
        "\t\t\t\tm_score = mat[i-1, j-1] + match\n",
        "\t\t\telse:\n",
        "\t\t\t\tm_score = mat[i-1, j-1] + mismatch\n",
        "\n",
        "\t\t\tchoices = [mat[i-1, j] + 1, mat[i, j-1] + 1, m_score]\n",
        "\t\t\tmat[i,j] = min(choices)\n",
        "\t\t\t#Traceback matrix\n",
        "\t\t\ttrace[i, j] = choices.index(min(choices)) \n",
        "\n",
        "\t#Traceback matrix\n",
        "\tpos0 = [seq1[-1]]\n",
        "\tpos1 = [seq2[-1]]\n",
        "\ti = -1 # index for pos0\n",
        "\tj = -1 # index for pos1\n",
        "\n",
        "\twhile (-i <= len(seq1)-1) & (-j <= len(seq2)-1):\n",
        "\t\tif trace[i, j] == 0:\n",
        "\t\t\tpos0.append('-')\n",
        "\t\t\tpos1.append(seq2[j])\n",
        "\t\t\ti -= 1\n",
        "\t\telif trace[i, j] == 1:\n",
        "\t\t\tpos0.append(seq1[i])\n",
        "\t\t\tpos1.append('-')\n",
        "\t\t\tj -= 1\n",
        "\t\telif trace[i, j] == 2:\n",
        "\t\t\tpos0.append(seq1[i-1])\n",
        "\t\t\tpos1.append(seq2[j-1])\n",
        "\t\t\ti -= 1\n",
        "\t\t\tj -=1\n",
        "\n",
        "\tpos0 = ''.join(pos0[::-1])\n",
        "\tpos1 = ''.join(pos1[::-1])\n",
        "\n",
        "\tprint(pos0,pos1)\n",
        "\n",
        "\t#count number of insertions or deletions\n",
        "\tdef count_events(p):\n",
        "\t\tres = re.finditer('-', p) #find locations with the gap\n",
        "\t\tres_start = [i.start() for i in res]\n",
        "\n",
        "\t\ti = 0\n",
        "\t\tevents = 0\n",
        "\n",
        "\t\twhile i < len(res_start):\n",
        "\t\t\tif i == 0: #if first nucleotide change\n",
        "\t\t\t\tevents += 1\n",
        "\t\t\telse:\n",
        "\t\t\t\tif res_start[i-1]+1 != res_start[i]: #if not consecutive\n",
        "\t\t\t\t\tevents += 1\n",
        "\t\t\ti+=1\n",
        "\t\treturn events\n",
        "\n",
        "\te0 = count_events(pos0)\n",
        "\te1 = count_events(pos1)\n",
        "\treturn e0, e1\n",
        "\n",
        "#Parsing the sequence file\n",
        "with open('/mnt/c/ds/input_data.txt') as f:\n",
        "\tseq_list = f.readlines()\n",
        "#seq_list = ['ATGC', 'ATGCGACCCC', 'ATGGGTATAT', 'TATATGATGC', 'GGGGCCCTCTATGCC', 'ATGTCTCTCCCC', 'ATTATATCTCTCGC', 'ATGCTCTCTCC', 'ATGTCCTCCC', 'GTGTGATGC', 'ATGCTCAGCTACGC', 'ATGCTCGACC']\n",
        "#seq_list = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k']\n",
        "seq_list.sort(key=len)\n",
        "\n",
        "bin_1 = seq_list[:20]\n",
        "bin_2 = seq_list[20:40]\n",
        "bin_3 = seq_list[40:60]\n",
        "bin_4 = seq_list[60:80]\n",
        "bin_5 = seq_list[80:]\n",
        "\n",
        "#seq_list = ['ATGC', 'ATGCC']\n",
        "#pair = []\n",
        "final_comb = []\n",
        "#pair_events_sum = []\n",
        "\n",
        "#Iterating throgh all bins\n",
        "for k in range(1,6):\n",
        "\tprint(k)\n",
        "\tbinname = locals()[\"bin_\"+str(k)]\n",
        "\tprint(\"this is binname\", str(k))\n",
        "\tfor r in range(len(binname) + 1):\n",
        "\t\tcombinations_object = itertools.combinations(binname, 2)\n",
        "\t\tcombinations_list = list(combinations_object)\n",
        "\t\tfinal_comb += combinations_list\n",
        "\t\tmyset = set(final_comb)\n",
        "\t\tfinal_comb = list(myset)\n",
        "\n",
        "\tfinal_comb_1 = [list(ele) for ele in final_comb]\n",
        "\t#print(\"Final comb\", final_comb_1)\n",
        "\n",
        "\n",
        "\tpair = []\n",
        "\tpair_events_sum = []\n",
        "#Finding combinations of 2 in all bins and aligning them\n",
        "\tfor i in final_comb:\n",
        "\t\t#print(\"this is i\", i)\n",
        "\t\tfor seq1, seq2 in zip(i[::2], i[1::2]):\n",
        "\t\t\tm = len(seq1) + 1 #seq1 length\n",
        "\t\t\tn = len(seq2) + 1 #seq2 length\n",
        "\t\t\tmat = np.zeros((m,n)) #scoring matrix of dimension mxn\n",
        "\t\t\ttrace = np.zeros((m,n)) #traceback matrix of dimension mxn\n",
        "\t\t\tevents0, events1 = aligner(seq1, seq2)\n",
        "\t\t\tsum = events0 + events1\n",
        "\t\t\t#print(\"Sum\", sum)\n",
        "\t\t\tpair_events_sum.append(sum)\n",
        "\t\t\t#print(\"This is pair_events_sum_list\", pair_events_sum)\n",
        "\t\t\tpair.append(i)\n",
        "\t\t\t#print(\"This is pair and sum of events\", pair, pair_events_sum)\n",
        "\tprint(\"finished aligning bin\", str(k))\n",
        "\tscores_df_bin = pd.DataFrame({'pair':pair, 'score':pair_events_sum})\n",
        "\tscores_df_bin.to_csv(\"Score for bin_\"+ str(k) + \".csv\", index = False)\n",
        "\n",
        "extension = 'csv'\n",
        "all_filenames = [i for i in glob.glob('*.{}'.format(extension))]\n",
        "#combine all files in the list\n",
        "combined_csv = pd.concat([pd.read_csv(f) for f in all_filenames ])\n",
        "#export to csv\n",
        "list_of_pairs = combined_csv['pair'].tolist()\n",
        "\n",
        "combined_csv.to_csv( \"Final score for all bins.csv\", index=False)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nbF99fzCegB5"
      },
      "source": [
        "Please note that this cell separates 2 different python files that I used for task 2"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "DvCjfC8FegB6"
      },
      "source": [
        "import pandas as pd\n",
        "import ast\n",
        "\n",
        "'''\n",
        "This code maps sequences to their number in order to output a file that has number n\n",
        "on i-th line, which means that n-th person was the i-th person to get the virus.\n",
        "'''\n",
        "def MapWordsToPos(dict_1,pair_list):\n",
        "    new_pair_list = []\n",
        "    for pair in pairlist:\n",
        "        new_l = []\n",
        "        #print(\"This is pair\", pair)\n",
        "        for seq in pair:\n",
        "            #print(\"This is seq\", seq)\n",
        "            if seq in seq_dict.keys():\n",
        "                val = seq_dict.get(seq)\n",
        "                #print(\"This is seq val\", seq,val)\n",
        "                new_l.append(val)\n",
        "                print(\"This is new list\", new_l)\n",
        "        new_pair_list.append(new_l)\n",
        "        print(\"This is new pair list\", new_pair_list)\n",
        "    return new_pair_list\n",
        "\n",
        "'''\n",
        "pairlist = [('a','b'), ('c', 'd'), ('e', 'f'), ('b', 'c')]\n",
        "seq_list = ['a', 'b', 'c', 'd', 'e', 'f']\n",
        "score = [1,3,2,4]\n",
        "'''\n",
        "\n",
        "df = pd.read_csv(\"/mnt/c/ds/Final score for all bins.csv\", converters={'pair': ast.literal_eval})\n",
        "print(df)\n",
        "pairlist = df['pair'].to_list()\n",
        "score = df['score'].to_list()\n",
        "#print(\"pairlist\", pairlist)\n",
        "#print(\"score\", score)\n",
        "\n",
        "with open('/mnt/c/ds/input_data.txt') as f:\n",
        "    seq_list = [i.strip('\\n') for i in f]\n",
        "print(seq_list)\n",
        "vals = list(range(1,101))\n",
        "seq_dict = {}\n",
        "\n",
        "for key in seq_list:\n",
        "    for value in vals:\n",
        "        seq_dict[key] = value\n",
        "        vals.remove(value)\n",
        "        break\n",
        "#print(\"Dict\", seq_dict)\n",
        "#print(\"Pair\", pairlist[0])\n",
        "\n",
        "\n",
        "new_pair_list = MapWordsToPos(seq_dict,seq_list)\n",
        "print(new_pair_list)\n",
        "\n",
        "\n",
        "d = {'Pair Indices':new_pair_list,'Score':score}\n",
        "df2 = pd.DataFrame(d)\n",
        "df2 = df2.sort_values(by=['Score'])\n",
        "#print(df2)\n",
        "\n",
        "pair_list = df2['Pair Indices'].to_list()\n",
        "order_list = []\n",
        "order_list.extend(pair_list[0])\n",
        "for pair in pair_list:\n",
        "    for seq in pair:\n",
        "        if seq in order_list:\n",
        "            continue\n",
        "        else:\n",
        "            order_list.append(seq)\n",
        "#print(order_list)\n",
        "\n",
        "\n",
        "textfile = open(\"task2_final_output.txt\", \"w\")\n",
        "for element in order_list:\n",
        "    textfile.write(str(element) + \"\\n\")\n",
        "textfile.close()"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "feG2KWuggPiM"
      },
      "source": [
        "The order is as below, I have attached the output file in the email\n",
        "\n",
        "36\n",
        " 34\n",
        " 68\n",
        " 83\n",
        " 26\n",
        " 6\n",
        " 16\n",
        " 56\n",
        " 88\n",
        " 50\n",
        " 40\n",
        " 51\n",
        " 43\n",
        " 86\n",
        " 77\n",
        " 13\n",
        " 70\n",
        " 25\n",
        " 53\n",
        " 49\n",
        " 22\n",
        " 8\n",
        " 15\n",
        " 41\n",
        " 58\n",
        " 93\n",
        " 4\n",
        " 69\n",
        " 63\n",
        " 44\n",
        " 82\n",
        " 30\n",
        " 39\n",
        " 46\n",
        " 12\n",
        " 80\n",
        " 59\n",
        " 92\n",
        " 37\n",
        " 95\n",
        " 90\n",
        " 84\n",
        " 47\n",
        " 76\n",
        " 99\n",
        " 54\n",
        " 73\n",
        " 61\n",
        " 9\n",
        " 67\n",
        " 57\n",
        " 3\n",
        " 28\n",
        " 71\n",
        " 87\n",
        " 100\n",
        " 78\n",
        " 14\n",
        " 27\n",
        " 24\n",
        " 74\n",
        " 65\n",
        " 75\n",
        " 45\n",
        " 7\n",
        " 23\n",
        " 48\n",
        " 17\n",
        " 38\n",
        " 98\n",
        " 96\n",
        " 97\n",
        " 20\n",
        " 64\n",
        " 21\n",
        " 60\n",
        " 52\n",
        " 33\n",
        " 29\n",
        " 2\n",
        " 19\n",
        " 55\n",
        " 79\n",
        " 62\n",
        " 10\n",
        " 31\n",
        " 85\n",
        " 89\n",
        " 5\n",
        " 72\n",
        " 18\n",
        " 91\n",
        " 1\n",
        " 66\n",
        " 32\n",
        " 35\n",
        " 11\n",
        " 94\n",
        " 81\n",
        " 42"
      ]
    }
  ]
}